types/evidence.go

```diff
@@ -245,6 +266,663 @@ func (dve *DuplicateVoteEvidence) ValidateBasic() error {
 	return nil
 }
 
+//-------------------------------------------
+
+// EvidenceList is a list of Evidence. Evidences is not a word.
+type EvidenceList []Evidence
+
+// Hash returns the simple merkle root hash of the EvidenceList.
+func (evl EvidenceList) Hash() []byte {
+	// These allocations are required because Evidence is not of type Bytes, and
+	// golang slices can't be typed cast. This shouldn't be a performance problem since
+	// the Evidence size is capped.
+	evidenceBzs := make([][]byte, len(evl))
+	for i := 0; i < len(evl); i++ {
+		evidenceBzs[i] = evl[i].Bytes()
+	}
+	return merkle.SimpleHashFromByteSlices(evidenceBzs)
+}
+
+func (evl EvidenceList) String() string {
+	s := ""
+	for _, e := range evl {
+		s += fmt.Sprintf("%s\t\t", e)
+	}
+	return s
+}
+
+// Has returns true if the evidence is in the EvidenceList.
+func (evl EvidenceList) Has(evidence Evidence) bool {
+	for _, ev := range evl {
+		if ev.Equal(evidence) {
+			return true
+		}
+	}
+	return false
+}
+
+//-------------------------------------------
+
+// ConflictingHeadersEvidence is primarily used by the light client when it
+// observes two conflicting headers, both having 1/3+ of the voting power of
+// the currently trusted validator set.
+type ConflictingHeadersEvidence struct {
+	H1 *SignedHeader `json:"h_1"`
+	H2 *SignedHeader `json:"h_2"`
+}
+
+var _ Evidence = &ConflictingHeadersEvidence{}
+var _ CompositeEvidence = &ConflictingHeadersEvidence{}
+var _ Evidence = ConflictingHeadersEvidence{}
+var _ CompositeEvidence = ConflictingHeadersEvidence{}
+
+// Split breaks up eviddence into smaller chunks (one per validator except for
+// PotentialAmnesiaEvidence): PhantomValidatorEvidence,
+// LunaticValidatorEvidence, DuplicateVoteEvidence and
+// PotentialAmnesiaEvidence.
+//
+// committedHeader - header at height H1.Height == H2.Height
+// valSet					 - validator set at height H1.Height == H2.Height
+// valToLastHeight - map between active validators and respective last heights
+func (ev ConflictingHeadersEvidence) Split(committedHeader *Header, valSet *ValidatorSet,
+	valToLastHeight map[string]int64) []Evidence {
+
+	evList := make([]Evidence, 0)
+
+	var alternativeHeader *SignedHeader
+	if bytes.Equal(committedHeader.Hash(), ev.H1.Hash()) {
+		alternativeHeader = ev.H2
+	} else {
+		alternativeHeader = ev.H1
+	}
+
+	// If there are signers(alternativeHeader) that are not part of
+	// validators(committedHeader), they misbehaved as they are signing protocol
+	// messages in heights they are not validators => immediately slashable
+	// (#F4).
+	for i, sig := range alternativeHeader.Commit.Signatures {
```

Comment:
The problem I see with detecting phantom validators is if we are in a fork and we don't realise yet (I know this is a very rare case) then an honest validator could be fooled into thinking that it is part of the validator set for this alternativeHeader and feels that it is perfectly entitled to voting. Now when the network partition stops and honest nodes reconnect they will instantly start forming phantom validator evidence against one another and both forks will slash the honest validators of the other fork

Reply:
In fact, unless a fork is detected at the root i.e. when the duplicate vote/amnesia attack happens it will almost certainly lead to a flurry of phantom validator evidence when it is discovered at a height that is greater than the root

Reply:
> it will almost certainly lead to a flurry of phantom validator evidence

I'd have to think about it more, but first thought - no, because 1/3+ can not change validator set in incorrect way (without buy-in from either 1/3s)

Reply:
But the validator set can still change in a correct way. As in over time, with different transactions submitted, the validator set could slowly change.

Reply:
let's maybe create a separate issue where we'll list all implications of full node following a fork, light client following a fork and what we can do about it.

Reply:
https://github.com/tendermint/tendermint/issues/4697